### PR TITLE
feat: add github funcs with github toolkit

### LINKED ISSUE #1077: [Feature Request]  Integrate github tools into github toolkit
### Required prerequisites

- [X] I have searched the [Issue Tracker](https://github.com/camel-ai/camel/issues) and [Discussions](https://github.com/camel-ai/camel/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [X] Consider asking first in a [Discussion](https://github.com/camel-ai/camel/discussions/new).

### Solution

1, add github toolkit of get all repo content;
2, add github toolkit of get PR content;
3, add github toolkit of get Issues content.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by raywhoelse**
## Description

Describe your changes in detail.

## Motivation and Context

`close #1077` 

- [ ] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [ ] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/toolkits/github_toolkit.py [hunk 1]
```diff
@@ -300,6 +300,135 @@ def create_pull_request(
...
+                user ID and the comment body.
+        """
+        # Retrieve the specific pull request
+        pr = self.repo.get_pull(number=pr_number)
+
+        # Collect the comments from the pull request
+        comments = []
+        for (
+            comment
+        ) in pr.get_comments():  # Returns all the comments in the pull request
```
Comment (WHALEEYE) [inline camel/toolkits/github_toolkit.py:404]:
```suggestion
        # Returns all the comments in the pull request
        for comment in pr.get_comments():  
```

File: camel/toolkits/github_toolkit.py [hunk 2]
```diff
@@ -177,9 +178,8 @@ def retrieve_issue_list(self) -> List[GithubIssue]:
                 title=issue.title,
                 body=issue.body,
                 number=issue.number,
-                file_path=issue.labels[
-                    0
-                ].name,  # we require file path to be the first label in the PR
+                # we require file path to be the first label in the PR
+                file_path=issue.labels[0].name,
                 file_content=self.retrieve_file_content(issue.labels[0].name),
```
Comment (WHALEEYE) [inline camel/toolkits/github_toolkit.py:183]: Although this is not related to this PR, but is there any reason we are implementing the file_path this way? @Wendong-Fan

**Review by WHALEEYE — COMMENTED** [review]
Thanks @raywhoelse, left some comments. By the way I think the newly added functions should also be added to `get_tools()`?

File: camel/toolkits/github_toolkit.py [hunk 3]
```diff
@@ -14,8 +14,9 @@
 
 import os
 from datetime import datetime, timedelta
-from typing import List, Optional
+from typing import Dict, List, Optional, Union
 
+from github.ContentFile import ContentFile
```
Comment (Wendong-Fan) [inline camel/toolkits/github_toolkit.py:19]: move this import with the class to make it optional

File: camel/toolkits/github_toolkit.py [hunk 4]
```diff
@@ -300,6 +300,135 @@ def create_pull_request(
         else:
             raise ValueError("PRs with multiple files aren't supported yet.")
 
+    def get_all_issues(self, state: str = "open") -> List[Dict[str, object]]:
+        r"""Retrieves all issues from the GitHub repository.
+
+        Args:
+            state (str): The state of issues to retrieve. Defaults to 'open'.
+            Options are 'open', 'closed', or 'all'.
```
Comment (Wendong-Fan) [inline camel/toolkits/github_toolkit.py:308]: docstring format

File: camel/toolkits/github_toolkit.py [hunk 5]
```diff
@@ -300,6 +300,135 @@ def create_pull_request(
...
+
+    def get_all_pull_requests(
+        self, state: str = "open"
+    ) -> List[Dict[str, object]]:
+        r"""Retrieves all pull requests from the GitHub repository.
+
+        Args:
+            state (str): The state of pull requests to retrieve. Defaults to
+            'open'.
+            Options are 'open', 'closed', or 'all'.
```
Comment (Wendong-Fan) [inline camel/toolkits/github_toolkit.py:345]: docstring format

File: camel/toolkits/github_toolkit.py [hunk 6]
```diff
@@ -300,6 +300,135 @@ def create_pull_request(
         else:
             raise ValueError("PRs with multiple files aren't supported yet.")
 
+    def get_all_issues(self, state: str = "open") -> List[Dict[str, object]]:
```
Comment (Wendong-Fan) [inline camel/toolkits/github_toolkit.py:303]:
```suggestion
    def get_issue_list(self, state: str = "open") -> List[Dict[str, object]]:
```

File: camel/toolkits/github_toolkit.py [hunk 7]
```diff
@@ -300,6 +300,135 @@ def create_pull_request(
...
+        Returns:
+            str: issues content details
+        """
+        try:
+            issue = self.repo.get_issue(number=issue_number)
+            return issue.body
+        except Exception as e:
+            return f"can't get Issue number {issue_number}: {e!s}"
+
+    def get_all_pull_requests(
```
Comment (Wendong-Fan) [inline camel/toolkits/github_toolkit.py:337]:
```suggestion
    def get_pull_request_list(
```

File: camel/toolkits/github_toolkit.py [hunk 8]
```diff
@@ -300,6 +300,135 @@ def create_pull_request(
...
+        return pull_requests_info
+
+    def get_pull_request_code(self, pr_number) -> List[Dict[str, str]]:
+        r"""Retrieves the code changes of a specific pull request.
+
+        Args:
+            pr_number (int): The number of the pull request to retrieve.
+
+        Returns:
+            list: A list of dictionaries where each dictionary contains the
```
Comment (Wendong-Fan) [inline camel/toolkits/github_toolkit.py:366]: same as below

```suggestion
            List[Dict[str, str]]: A list of dictionaries where each dictionary contains the
```

File: camel/toolkits/github_toolkit.py [hunk 9]
```diff
@@ -300,6 +300,135 @@ def create_pull_request(
...
+                if content.type == "dir":
+                    # If it's a directory, recursively retrieve its file paths
+                    files.extend(self.get_all_file_paths(content.path))
+                else:
+                    # If it's a file, add its path to the list
+                    files.append(content.path)
+        return files
+
     def retrieve_file_content(self, file_path: str) -> str:
         r"""Retrieves the content of a file from the GitHub repository.
```
Comment (Wendong-Fan) [inline camel/toolkits/github_toolkit.py:434]: can we further add support for multiple files?

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @raywhoelse , I think `def get_tools(self)` also need to be updated, we can put it in the bottom, there are some redundant methods like `retrieve_issue_list` with `get_all_issues` and `retrieve_pull_requests` with `get_all_pull_requests` could you further clean the code?

**Comment by raywhoelse** [comment]
> Thanks @raywhoelse , I think `def get_tools(self)` also need to be updated, we can put it in the bottom, there are some redundant methods like `retrieve_issue_list` with `get_all_issues` and `retrieve_pull_requests` with `get_all_pull_requests` could you further clean the code?

yes, I will be updated, and thanks for will suggestions and comments, and also thanks for wendong comments. @WHALEEYE @Wendong-Fan

File: camel/toolkits/github_toolkit.py (hunk 2 above)
Comment (Wendong-Fan) [reply camel/toolkits/github_toolkit.py]: this has been removed

**Review by Wendong-Fan — APPROVED** [review]
thanks @raywhoelse , added one more commit here: https://github.com/camel-ai/camel/pull/1099/commits/aa63872c2e71a270d070a97e049777401aade802

**Pull request merged**